File: strategies/rsi-volitality-p1/rsi-volitality-p1.py
import numpy as np
import pandas as pd
import requests
from typing import Dict, List, Any

class RSIVolatilityStrategy:

	def __init__(self):
		# Strategy Parameters
		self.rsi_length = 14
		self.vol_length = 20
		self.rsi_overbought = 70
		self.rsi_oversold = 30
		self.vol_threshold = 1.5
		
		# Visualization Parameters
		self.colors = {
			'rsi_line': '#2196F3',  # Blue
			'overbought_line': '#FF5252',  # Red
			'oversold_line': '#4CAF50',  # Green
			'vol_high': '#FF9800',  # Orange
			'vol_normal': '#9E9E9E',  # Gray
			'signal_long': '#4CAF50',  # Green
			'signal_short': '#FF5252'  # Red
		}
		self.line_widths = {
			'rsi_line': 2,
			'level_lines': 1,
			'vol_line': 2
		}

	def calculate_rsi(self, closes: np.ndarray) -> np.ndarray:
		"""Calculate RSI indicator"""
		delta = np.diff(closes)
		gain = np.where(delta > 0, delta, 0)
		loss = np.where(delta < 0, -delta, 0)
		
		# Prepend zero to match array length
		gain = np.insert(gain, 0, 0)
		loss = np.insert(loss, 0, 0)
		
		# Calculate average gain and loss
		avg_gain = pd.Series(gain).rolling(window=self.rsi_length).mean().values
		avg_loss = pd.Series(loss).rolling(window=self.rsi_length).mean().values
		
		# Calculate RS and RSI
		rs = np.divide(avg_gain, avg_loss, out=np.zeros_like(avg_gain), where=avg_loss != 0)
		rsi = 100 - (100 / (1 + rs))
		
		return rsi

	def calculate_volatility(self, closes: np.ndarray) -> np.ndarray:
		"""Calculate price volatility"""
		returns = np.diff(np.log(closes))
		returns = np.insert(returns, 0, 0)
		volatility = pd.Series(returns).rolling(window=self.vol_length).std().values
		return volatility

	async def fetch_candle_data(self, symbol: str, timeframe: str) -> Dict[str, List[float]]:
		"""Fetch candle data from the API"""
		try:
			response = requests.get(
				f'http://localhost:5000/fetch_candles',
				params={'symbol': symbol, 'timeframe': timeframe}
			)
			response.raise_for_status()
			data = response.json()
			
			return {
				'open': [candle['open'] for candle in data],
				'high': [candle['high'] for candle in data],
				'low': [candle['low'] for candle in data],
				'close': [candle['close'] for candle in data],
				'volume': [candle.get('volume', 0) for candle in data],
				'time': [candle['time'] for candle in data]
			}
		except Exception as e:
			print(f"Error fetching candle data: {e}")
			return {'open': [], 'high': [], 'low': [], 'close': [], 'volume': [], 'time': []}

	def generate_signals(self, rsi: np.ndarray, volatility: np.ndarray, 
						vol_sma: np.ndarray) -> tuple[List[bool], List[bool]]:
		"""Generate trading signals based on RSI and volatility"""
		long_signals = []
		short_signals = []
		
		for i in range(len(rsi)):
			# High volatility condition
			high_vol = volatility[i] > (vol_sma[i] * self.vol_threshold)
			
			# Generate signals only in high volatility conditions
			if high_vol:
				long_signals.append(rsi[i] < self.rsi_oversold)
				short_signals.append(rsi[i] > self.rsi_overbought)
			else:
				long_signals.append(False)
				short_signals.append(False)
		
		return long_signals, short_signals

	def calculate_strategy(self, data: Dict[str, List[float]]) -> Dict[str, Any]:
		"""Main calculation function"""
		closes = np.array(data['close'])
		
		# Calculate indicators
		rsi = self.calculate_rsi(closes)
		volatility = self.calculate_volatility(closes)
		vol_sma = pd.Series(volatility).rolling(window=self.vol_length).mean().values
		
		# Generate signals
		long_signals, short_signals = self.generate_signals(rsi, volatility, vol_sma)
		
		return {
			'rsi': rsi.tolist(),
			'volatility': volatility.tolist(),
			'vol_sma': vol_sma.tolist(),
			'long_signals': long_signals,
			'short_signals': short_signals,
			'timestamps': data['time'],
			'visualization': {
				'colors': self.colors,
				'line_widths': self.line_widths
			}
		}

def calculate_indicators(symbol: str, timeframe: str) -> Dict[str, Any]:
	"""Main calculation function exposed to API"""
	strategy = RSIVolatilityStrategy()
	
	# Fetch candle data
	data = strategy.fetch_candle_data(symbol, timeframe)
	if not data['close']:
		return {
			'rsi': [],
			'volatility': [],
			'vol_sma': [],
			'long_signals': [],
			'short_signals': [],
			'timestamps': [],
			'visualization': {
				'colors': strategy.colors,
				'line_widths': strategy.line_widths
			}
		}
	
	# Calculate strategy indicators and signals
	result = strategy.calculate_strategy(data)
	return result

if __name__ == "__main__":
	print("RSI Volatility Strategy Module")

File: strategies/rsi-volitality-p1/rsi-volitality-p1.ts
import { LineStyle, LineStyleType } from 'lightweight-charts';

interface StrategyResult {
    rsi: number[];
    volatility: number[];
    vol_sma: number[];
    long_signals: boolean[];
    short_signals: boolean[];
    timestamps: number[];
    visualization: {
        colors: {
            rsi_line: string;
            overbought_line: string;
            oversold_line: string;
            vol_high: string;
            vol_normal: string;
            signal_long: string;
            signal_short: string;
        };
        line_widths: {
            rsi_line: number;
            level_lines: number;
            vol_line: number;
        };
    };
}

export class RSIVolatilityChart {
    private chart: any;
    private rsiSeries: any;
    private volatilitySeries: any;
    private volSmaSeries: any;
    private signalMarkers: any;
    private overboughtLine: any;
    private oversoldLine: any;

    constructor(container: HTMLElement) {
        // Initialize main chart
        this.chart = createChart(container, {
            layout: {
                background: { color: '#ffffff' },
                textColor: '#333',
            },
            grid: {
                vertLines: { color: '#f0f0f0' },
                horzLines: { color: '#f0f0f0' },
            },
            rightPriceScale: {
                borderColor: '#d1d4dc',
            },
            timeScale: {
                borderColor: '#d1d4dc',
                timeVisible: true,
            },
            width: container.clientWidth,
            height: container.clientHeight,
        });

        // Initialize RSI series
        this.rsiSeries = this.chart.addLineSeries({
            title: 'RSI',
            color: '#2196F3',
            lineWidth: 2,
            priceFormat: {
                type: 'custom',
                minMove: 0.1,
                formatter: (price: number) => price.toFixed(1),
            },
        });

        // Initialize Volatility series
        this.volatilitySeries = this.chart.addLineSeries({
            title: 'Volatility',
            color: '#FF9800',
            lineWidth: 2,
            priceFormat: {
                type: 'custom',
                minMove: 0.0001,
                formatter: (price: number) => price.toFixed(4),
            },
        });

        // Initialize Volatility SMA series
        this.volSmaSeries = this.chart.addLineSeries({
            title: 'Vol SMA',
            color: '#9E9E9E',
            lineWidth: 1,
            lineStyle: LineStyle.Dotted,
            priceFormat: {
                type: 'custom',
                minMove: 0.0001,
                formatter: (price: number) => price.toFixed(4),
            },
        });

        // Initialize signal markers series
        this.signalMarkers = this.chart.addLineSeries({
            title: 'Signals',
            lineVisible: false,
            lastValueVisible: false,
            priceFormat: {
                type: 'custom',
                minMove: 0.1,
                formatter: (price: number) => price.toFixed(1),
            },
        });

        // Initialize overbought and oversold lines
        this.overboughtLine = this.chart.addLineSeries({
            title: 'Overbought',
            color: '#FF5252',
            lineWidth: 1,
            lineStyle: LineStyle.Dashed,
        });

        this.oversoldLine = this.chart.addLineSeries({
            title: 'Oversold',
            color: '#4CAF50',
            lineWidth: 1,
            lineStyle: LineStyle.Dashed,
        });
    }

    public async updateChart(symbol: string, timeframe: string): Promise<void> {
        try {
            // Use Next.js API route
            const response = await fetch(`/api/rsi-volatility?symbol=${symbol}&timeframe=${timeframe}`);
            if (!response.ok) {
                throw new Error('Failed to fetch data');
            }
            const result: StrategyResult = await response.json();
            
            // Prepare data series
            const rsiData = result.rsi.map((value, index) => ({
                time: result.timestamps[index] as number,
                value: value,
            }));

            const volatilityData = result.volatility.map((value, index) => ({
                time: result.timestamps[index] as number,
                value: value,
            }));

            const volSmaData = result.vol_sma.map((value, index) => ({
                time: result.timestamps[index] as number,
                value: value,
            }));

            // Prepare signal markers
            const markers = result.timestamps.map((time, index) => {
                if (result.long_signals[index] || result.short_signals[index]) {
                    return {
                        time: time as number,
                        position: result.long_signals[index] ? 'belowBar' : 'aboveBar',
                        color: result.long_signals[index] ? 
                            result.visualization.colors.signal_long : 
                            result.visualization.colors.signal_short,
                        shape: result.long_signals[index] ? 'arrowUp' : 'arrowDown',
                        text: result.long_signals[index] ? 'BUY' : 'SELL',
                    };
                }
                return null;
            }).filter(marker => marker !== null);

            // Update all series
            this.rsiSeries.setData(rsiData);
            this.volatilitySeries.setData(volatilityData);
            this.volSmaSeries.setData(volSmaData);
            this.signalMarkers.setMarkers(markers);

            // Set overbought and oversold lines
            const timeRange = {
                from: Math.min(...result.timestamps),
                to: Math.max(...result.timestamps),
            };

            this.overboughtLine.setData([
                { time: timeRange.from, value: 70 },
                { time: timeRange.to, value: 70 },
            ]);

            this.oversoldLine.setData([
                { time: timeRange.from, value: 30 },
                { time: timeRange.to, value: 30 },
            ]);

            // Fit content
            this.chart.timeScale().fitContent();

        } catch (error) {
            console.error('Error updating chart:', error);
        }
    }

    public resize(width: number, height: number): void {
        this.chart.resize(width, height);
    }

    public remove(): void {
        this.chart.remove();
    }
}